stdlib/iarray.mli

```diff
@@ -98,111 +118,319 @@ val of_array : 'a array -> 'a iarray
 
 (** {1 Iterators} *)
 
-val iter : ('a -> unit) -> 'a iarray -> unit
+val iter : local_ ('a -> unit) -> 'a iarray -> unit
 (** [iter f a] applies function [f] in turn to all
    the elements of [a].  It is equivalent to
    [f a.:(0); f a.:(1); ...; f a.:(length a - 1); ()]. *)
 
-val iteri : (int -> 'a -> unit) -> 'a iarray -> unit
+val iter_local : local_ (local_ 'a -> unit) -> local_ 'a iarray -> unit
+(** The locally-scoped version of [iter]. *)
+
+val iteri : local_ (int -> 'a -> unit) -> 'a iarray -> unit
 (** Same as {!iter}, but the
    function is applied to the index of the element as first argument,
    and the element itself as second argument. *)
 
-val map : ('a -> 'b) -> 'a iarray -> 'b iarray
+val iteri_local : local_ (int -> local_ 'a -> unit) -> local_ 'a iarray -> unit
+(** The locally-scoped version of [iteri]. *)
+
+val map : local_ ('a -> 'b) -> 'a iarray -> 'b iarray
 (** [map f a] applies function [f] to all the elements of [a],
    and builds an immutable array with the results returned by [f]:
    [[| f a.:(0); f a.:(1); ...; f a.:(length a - 1) |]]. *)
 
-val mapi : (int -> 'a -> 'b) -> 'a iarray -> 'b iarray
+val map_local :
+  local_ (local_ 'a -> local_ 'b) -> local_ 'a iarray -> local_ 'b iarray
+(** The locally-scoped and locally-allocating version of [map]. *)
+
+val map_local_input : local_ (local_ 'a -> 'b) -> local_ 'a iarray -> 'b iarray
+(** The locally-constrained but globally-allocating version of [map]. *)
+
+val map_local_output : local_ ('a -> local_ 'b) -> 'a iarray -> local_ 'b iarray
+(** The locally-allocating but global-input version of [map]. *)
+
+val mapi : local_ (int -> 'a -> 'b) -> 'a iarray -> 'b iarray
 (** Same as {!map}, but the
    function is applied to the index of the element as first argument,
    and the element itself as second argument. *)
 
-val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b iarray -> 'a
+val mapi_local :
+  local_ (int -> local_ 'a -> local_ 'b) ->
+  local_ 'a iarray ->
+  local_ 'b iarray
+(** The locally-scoped and locally-allocating version of [mapi]. *)
+
+val mapi_local_input :
+  local_ (int -> local_ 'a -> 'b) -> local_ 'a iarray -> 'b iarray
+(** The locally-constrained but globally-allocating version of [mapi]. *)
+
+val mapi_local_output :
+  local_ (int -> 'a -> local_ 'b) -> 'a iarray -> local_ 'b iarray
+(** The locally-allocating but global-input version of [mapi]. *)
+
+val fold_left : local_ ('a -> 'b -> 'a) -> 'a -> 'b iarray -> 'a
 (** [fold_left f init a] computes
    [f (... (f (f init a.:(0)) a.:(1)) ...) a.:(n-1)],
    where [n] is the length of the immutable array [a]. *)
 
+(* MISSING: Requires [let mutable]
```

Comment:
Really? I don't see why.

Reply:
If you look at the implementation in `array.ml`, you'll see that it uses a locally-scoped `ref`; we can't put local values in there, which is exactly what `let mutable` solves.

Reply:
Well, yes, but it looks possible to avoid the `ref`:

```
let fold_left_local :
  local_ (local_ 'a -> local_ 'b -> local_ 'a) ->
  local_ 'a ->
  local_ 'b iarray ->
  local_ 'a =
  fun f z arr -> local_
  let len = length arr in
  let rec go acc i = local_
    if i = len
    then acc
    else go (f acc (unsafe_get arr i)) (i+1)
  in
  go z 0
```

This is still not that great, because it permanently consumes O(n) stack space (well, until the outer call's region ends). But I don't see a way to avoid the stack consumption, because the final value must go *somewhere*. By induction, all previous values must go *somewhere*. And each one can't be gotten rid of until the next one is computed. So unless the function gets some kind of `copy:('a -> 'a)` argument, I don't see a way to avoid the stack usage. Maybe someone else can.

Note that the `let mutable` solution is no better in this regard: each intermediate `'a` would end up sitting on the stack until the outer caller's region is popped.

Do add a comment in the mli file warning users that all their intermediate values will be put on the stack. In the end, they probably don't want this function, but it's not for us to tell users what they do and don't want.